Intellisense fix for Visual Studio 15.3 and up

francipvb opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:

Closes #8336
Closes #9896 

### Summary of the issue:

When Intellisense popup is shown in Visual Studio and NVDA reads the item, the current line is also read, taking some time before the suggestion is read out.

### Description of how this pull request fixes the issue:

I've tried a more complex approach, but I ended with a speech stop before live region change event.

### Testing performed:

Tested against latest update of Visual Studio 2017 and latest preview of 2019.

### Known issues with pull request:

None known.

### Change log entry:

Section: Bug fixes
* In Visual Studio 2017 (15.3) and up, speech is stopped if an Intellisense popup is open.

EDIT: Changed issue link.

francipvb committed:
Changes:

* Added `major` and `minor` properties to the appModule.
* Added overlay classes for the intellisense item and the intellisense label.
  - The event `liveRegionChange` was overridden to stop the speech at time of speak.

francipvb commented:
CC @leonardder @josephsl @Adriani90 

Sorry but I don't know how to roll back to a previous version of Visual Studio without breaking the Visual Studio installation.

Cheers,

francipvb committed:
Fixed flake8 issues.

leonardder commented on source/appModules/devenv.py:
```diff
@@ -89,9 +92,14 @@
 class AppModule(appModuleHandler.AppModule):
+
+	def _get_major(self):
+		return int(self.productVersion.split(".", 2)[0])
```
Rather than having separate attributes for major and minor, could you please make this an attribute that returns a tuple? This can be similar to the code you removed from chooseNVDAObjectOverlayClasses

francipvb replied:
Done, I will upload a change now.

leonardder commented on source/appModules/devenv.py:
```diff
@@ -102,11 +110,30 @@ def chooseNVDAObjectOverlayClasses(self, obj, clsList):
+
+			try:
+				if (
+					obj.role == controlTypes.ROLE_STATICTEXT
```
I"m afraid that this code will have major impact on performance, as it will create several UIA objects for objects we do not use.

francipvb replied:
How can I avoid this issue?

Thanks,

francipvb committed:
Changes:

* Removed IntellisenseItem class.
* Changed the check for the first child of Intellisense list to the list itself.
* Added a property that returns a tuple insthead the old two returning major and minor versions.

leonardder commented on source/appModules/devenv.py:
```diff
@@ -102,11 +114,25 @@ def chooseNVDAObjectOverlayClasses(self, obj, clsList):
+				if (
+					obj.role == controlTypes.ROLE_STATICTEXT
+					and isinstance(obj, UIA.UIA)
+					and obj.previous.previous.UIAElement.CachedAutomationId in self.INTELLISENSE_LIST_AUTOMATIONIDS
```
```suggestion
					and obj.previous
					and obj.previous.previous
					and obj.previous.previous.UIAElement.CachedAutomationId in self.INTELLISENSE_LIST_AUTOMATIONIDS
```

Unfortunately, Python has no such way as in C# to use something like
```
					and obj.previous?.previous?.UIAElement?.CachedAutomationId in self.INTELLISENSE_LIST_AUTOMATIONIDS
```

francipvb replied:
Thanks @leonardder, this makes sense.

Cheers,

francipvb committed:
Added checks for "None" values in references to siblings.

This may improve performance and avoid "AttributeError" exceptions.

Thanks to Leonard for this change suggestion.

Co-Authored-By: Leonard de Ruijter <leonardder@users.noreply.github.com>

francipvb committed:
Switched to 'item status event' approach for speaking the current selected item.

francipvb committed:
Merge branch 'master' into i9440

francipvb committed:
Fixed flake8 and reference errors.

leonardder commented:
@francipvb Would you mind if I file a pr against your fork to implement this in an alternative way that also works for SSMS?

francipvb commented:
Hello,

I have no problem. Please take a look to #10394 too if you can do so.

Cheers,

leonardder commented:
It turns out that the uia_elementSelected event is also fired on the selected intellisense item. I think I prefer using that, as it doesn't require us to fetch item status. I will also try to implement suggestion announcement with suggestion sounds.
